pkg/server/autoconfig/doc.go

```diff
@@ -0,0 +1,106 @@
+//     (see below for details), and organizes their execution
+//     by creating individual 'auto config task' jobs.
+//   - Each auto config task job runs one task; it is cancellable
+//     by the end user.
+//   - The runner job only creates the job for the next task
+//     after the job for the previous task has fully completed
+//     (either successfully, with an error, or has been cancelled).
+//
+// # Task provenance
+//
+// The auto config runner job takes its input from a "task provider"
+// (Go interface `Provider`, from the package `acprovider`). The task
+// provider has a notification channel and an accessor that lists
+// pending tasks.
+//
+// The provider also has a method through which the task runner can
+// advise of task completion. This is an optimization and can be used
+// by the provider to truncate the list of incoming tasks.
+//
+// In practice, there are 2 main task providers envisioned at the time
+// of this writing:
+//
+//   - A provider of static configuration tasks for the system tenant,
+//     via hardcoded "configuration profiles" selected when a cluster is
+//     initially created.
+//   - A provider of dynamic configuration asks for secondary tenants,
+//     provisioned through a RPC in the "tenant connector".
+//
+// # Task definition and execution guarantees
+//
+// The runner logic provides at-least-once execution semantics, and
+// tries hard to provide exactly-once semantics. (As the theory tells
+// us that exactly-once is not possible, see below for a discussion of
+// exceptions.) It also guarantees that auto config tasks are not
+// executed concurrently with each other, with at most one node
+// executing a task at a time.
+//
+// In a nutshell, the mechanism works as follows:
+//
+//   - Tasks are numbered sequentially (task ID); it is the
+//     responsibility of "the environment" (the task provider) to
+//     guarantee that later tasks have greater IDs than earlier tasks.
+//   - The runner uses "started" and "completed" markers in the
+//     system.job_info table to ensure that no more than one job is
+//     created for a given task.
+//
+// More specifically, when the runner is ready to execute a task, it
+// transactionally:
+//
+//  1. checks that there is no in-flight task already.
+//     (e.g. one started by the runner on another node)
+//  2. retrieves the latest completion marker by task ID from job_info,
+//     (indicates the latest task completed by ID)
+//  3. it ratchets the queue of incoming tasks forward
+//     to skip over all tasks with an ID lower or equal to the
+//     last task completed.
+//  4. it creates a "auto config task" job entry for that task
+//     and places a start marker for it in job_info.
+//
+// Then, the "auto config task" executes the SQL in the task
+// transactionally with the removal of its start marker and the
+// creation of its completion marker.
+//
+// # Possible exceptions to the "exactly-once" semantics
```

Comment:
It seems to me that job success is at most once. If these tasks fail with a permanent error, they are not retried by anything. Cancellation is a mechanism to cause failure. One thing we might want to expose is a way to mark a task as required. In jobs parlance, this would mean that it's non-cancelable. A non-cancelable job treats errors as retryable by default (as opposed to permanent), and it prevents user cancelation. I suspect for the hard-coded tasks, we'll want to mark them as uncancelable.